### DISCUSSION TITLE
Adding custom dataset

### DISCUSSION THREAD

**Discussion by AlokRaj1994** [deepdoctection/deepdoctection#42] [Q&A] [answered]
How can we add a new dataset to the the existing list of dataset?

**Comment by JaMe76** [comment, marked as answer]
Hi, 
a small tutorial about writing a custom dataset can be found in the [docs](https://deepdoctection.readthedocs.io/en/latest/tutorials/datasets.html) . An update of this tutorial can be found [here](https://github.com/deepdoctection/deepdoctection/blob/master/docs/tutorials/datasets.rst) . The updates will appear in the docs the next release.  

Feel free to ask, if there is anything not clear to you.

**Reply by AlokRaj1994** [reply]
It will be useful if you add a notebook explaining creation of new dataset containing coco format annotations and one category.

**Reply by JaMe76** [reply]
I see, I  will think about that.

Right now, regarding coco format annotations I suggest looking at the implementation of the Publaynet dataset. It has a coco format annotation file.

**Reply by AlokRaj1994** [reply]
I can see a file publaynet.py file inside instances folder which contains classes publaynet and publaynetbuilder like in documentation of custom dataset ,so should i create that kind of a file for new dataset?

**Reply by JaMe76** [reply]
Yeah, using code there snippets might be easier to start with. 

Ignore the handling of the optional parameters and try to implement the # Load and # Map blocks. I also suggest to ignore everything below line 131 (apart from the return value of course). If your dataset is  coco style, this will get you quite far.

**Reply by AlokRaj1994** [reply]
I should put the dataset with its annotation(.json) file in  .cache directory of deepdoctection right?

**Reply by JaMe76** [reply]
yes

**Reply by AlokRaj1994** [reply]
How can i register the dataset?

**Reply by JaMe76** [reply]
Registering custom dataset is something that does work as it should right now. At least I could not figure out how to make it work smoothly. 

Not having the registry at hand, you can just instantiate the custom_dataset from your customdataset.py like

```
from customdataset import  custom_dataset 

custom_instance = custom_dataset()
```

You don't need to have to register your dataset. The registry simplifies instantiating a dataset so that you can call 

```
import deepdoctection as dd

custom_dataset = dd.get_dataset("rvl") 

df = custom_dataset.dataflow.build()

for dp in df:
 ...
```

**Reply by AlokRaj1994** [reply]
After fine tune how can we use the model to predict ?
Can we use the model in get started notebook?

**Reply by JaMe76** [reply]
You can use it in pipeline components or pipelines where the model has been used before. You just need to replace the old model weights with the new model weights in configs.

**Reply by AlokRaj1994** [reply]
Getting an error that model is not registered
Model is not registered. Please make sure the weights are available in the weights cache directory

**Reply by AlokRaj1994** [reply]
Actually i placed that model file in .cache directory still its showing error

**Reply by JaMe76** [reply]
This is info only. You must save weights in subdirectory /.cache/deepdoctection/weights..., depending what model you fine-tuned.

**Reply by AlokRaj1994** [reply]
after fine tuning i got a file which i placed in .cache/deepdoctection/weights 
The info_only and publaynet model downloaded from HF is working but the fine-tuned model not working
All of these 3 files are in .cache/deepdoctection/weights.

**Reply by JaMe76** [reply]
Files must be placed into one of the sub directories of weights: /layout, /cell or /item depending on what model you train. 

Also number of files you need depend on framework you train: D2 has .pkl file (e.g. d2_model-1800000-cell.pkl), Tensorflow need two files:
(e.g. model-1800000.data-00000-of-00001 and model-1800000.index).

**Reply by AlokRaj1994** [reply]
yes i put that under cell
2 files similar to above

**Reply by JaMe76** [reply]
I then need the error message. 

As mentioned above not registered models is info only and will not break the code.

**Reply by AlokRaj1994** [reply]
This was the error 
![dp_1](https://user-images.githubusercontent.com/66828266/183343387-6c1da9e6-3ab6-4184-83e7-c4169d6dadce.jpg)

**Reply by JaMe76** [reply]
I see, you need to register the fine tuned model. This was originally not intended and has to be fixed. Will open an issue for that. 

As work around for now you have to register the new weights, e.g.

```
    from deepdoctection.extern.model import ModelCatalog, ModelProfile

    ModelCatalog.register("cell/model-1810000_inf_only.data-00000-of-00001",ModelProfile(
        name="cell/model-1810000_inf_only.data-00000-of-00001",
        description="Tensorpack cell detection model for inference purposes trained on Pubtabnet",
        config="dd/tp/conf_frcnn_cell.yaml",
        size=[274503056, 8056],
        tp_model=True,
        categories={"1": "CELL"}))
```

Add the file size of model weights model-1800000.data-00000-of-00001 and model-1800000.index in ```ModelProfile.size```. 
You can get the number using:

```
    import os

    stat_info = os.stat("path/to/dir")
    size = stat_info.st_size
```

**Reply by AlokRaj1994** [reply]
I have registered it but still it is not listed when i run the below code
print(ModelCatalog.get_model_list())

**Reply by JaMe76** [reply]
Does it show up in ModelCatalog.CATALOG ? 

Note, that newly registered models are currently only available at run time. There is no physical saving of the profile. Once you restart notebook, you have to re-register. That's the reason why I want to avoid other using the registry.

**Reply by AlokRaj1994** [reply]
It was not shown in ModelCatalog.CATALOG.

**Reply by AlokRaj1994** [reply]
Now it is registered and showing in ModelCatalog.CATALOG
Thank you

**Reply by AlokRaj1994** [reply]
If we are using a custom dataset what are the required objects in annotation file.
Like filename,bbox etc
Are html and tokens required?
And can you please help with some more documentation to setup and register custom dataset

**Reply by JaMe76** [reply]
For now, I would recommend to look into the coco mapping function, as this dataset has the minimal object detection requirements. 

Pubtabnet can also be used for object detection, but for many other things as well (e.g. table structure recognition, row, column object detection). This is the reason why the Pubtabnet mapping function is a lot more complex and a lot more will be mapped from the annotation file. 

With respect to extending the documentation for setting up and registering datasets, i want to rely more on concrete questions for now (and maybe writing an easier class for setting up custom datasets for specific use cases). 
At this point, however, I am not quite sure how to extend documentation in order to be sure that more advise will actually mean to be more helpful.

**Reply by AlokRaj1994** [reply]
I have put the file(new dataset containing images and json file) to /.cache/deepdoctection/datasets
Now i should create customdataset.py file in instances right?
Then to register it what should i do?
Or how can i use it in fine-tuning ?

**Reply by AlokRaj1994** [reply]
Followed the steps in documentation.
But showing some error while building dataflow.
![dataflow_build](https://user-images.githubusercontent.com/66828266/183822733-3492affb-8b53-46f0-9884-4f8227da31a6.jpg)

**Reply by JaMe76** [reply]
Error is StopIteration which indicates that iteration has been coming to its end. It is likely that your dataflow pipeline is clogged at some step
I suggest to go through your `build` method and see if at any step your datapoints gets filtered, lost etc. 

Example, how error pops up. Last step shows StopIteration
```
   import deepdoctection as dd

   A = [1,2]
   df = dd.DataFromList(A)
   a_iter = iter(df)
   print(next(a_iter))
   print(next(a_iter))
   print(next(a_iter))
```

**Reply by AlokRaj1994** [reply]
![build_1](https://user-images.githubusercontent.com/66828266/183851270-03bbb5f8-bbf4-40bd-bfee-9777f14488ee.jpg)
Is this enough for a dataset with annotation in coco format

**Reply by AlokRaj1994** [reply]
why this error comes?
[0810 14:51.45 @logger.py:201] INF dp: /home/alokraj/.cache/deepdoctection/datasets/invoice/train/994c4ea5-ac4c4c8_0625smart.jpg, err: KeyError,
            msg: '6' in: <frame at 0x6b1edf0, file '/home/alokraj/.local/lib/python3.8/site-packages/deepdoctection/mapper/cocostruct.py', line 89, code coco_to_image> will be filtered

**Reply by JaMe76** [reply]
Between these lines in coco_to_image you get a KeyError. 

![coco_to_image](https://user-images.githubusercontent.com/50324223/183884546-1477f4c1-0311-4730-9314-2831af67f481.png)

MappingContextManager catches KeyError, ValueError, IndexError, AssertionError, TypeError and jumps to  

```
if mapping_context.context_error:
        return None
```

**Reply by JaMe76** [reply]
Seems like there is no key like `file_name`, `category_id` or `id` in your annotation file.

**Reply by AlokRaj1994** [reply]
![ann_1](https://user-images.githubusercontent.com/66828266/183890587-90d7cef5-edce-4f18-9efa-ceb87db1f6d3.jpg)
annotation

**Reply by JaMe76** [reply]
That's only annotation, coco format also has `image` and `categories`.  

Check what going on in your `build` method:

``` 
df = SerializerCoco.load(path_annotation_file)

dp = next(iter(df))
```
SerializerCoco assembles everything from image and annotations to create and stream complete for one single datapoint. Output `dp` must contain all keys used in `coco_to_image` mapping.

**Reply by AlokRaj1994** [reply]
![coco_1](https://user-images.githubusercontent.com/66828266/184068085-85e44a23-f2b3-4c13-b505-9ca93d455b97.jpg)
This was executed without error.

**Reply by JaMe76** [reply]
Yes and did you look into output of  

```
dp = next(iter(df))
```
? There will be something missing probably.

**Reply by AlokRaj1994** [reply]
![coco_3](https://user-images.githubusercontent.com/66828266/184075428-ab18c740-d4f8-411f-a2b2-e44f0d45829f.jpg)
output

**Reply by JaMe76** [reply]
Looks good. 

What is returned in 

```
self.categories.get_categories()
```
?

It must be a dict mapping category_ids to category names. ids in dict must be strings, e.g.

```
self.categories.get_categories() = {"1": "TEXT", "2": "TITLE"}
```

**Reply by AlokRaj1994** [reply]
That's right as the same id in annotation file

**Reply by AlokRaj1994** [reply]
My categories id start from 0 but  self.categories.get_categories() start from 1
should i change the annotation file?

**Reply by AlokRaj1994** [reply]
I have changed it but still its not working

**Reply by JaMe76** [reply]
0 is reserved for background in model. Category id for objects in datasets must start with 1. Increasing the id in annotation file by 1 is the simplest way i think.

**Reply by AlokRaj1994** [reply]
yes i increased but still not working

**Reply by JaMe76** [reply]
I would write a simple mapping and plug it the dataflow build stream. You do not need to change the annotation file physically. Will look something like this.

``` 

        df = SerializerCoco.load(path_annotation_file)

        def increment_category_id(dp: JsonDict) -> JsonDict:
            # here you increase id by 1
            return dp

        df = MapData(df, increment_category_id)
        
       
        coco_to_image_mapper = ...
        df = MapData(df, coco_to_image_mapper)
        
```

**Reply by AlokRaj1994** [reply]
With code i incremented the id values but still its showing stop iteration error

**Reply by AlokRaj1994** [reply]
Fine now i got it
Thank you

**Reply by AlokRaj1994** [reply]
UserWarning: Converting sparse IndexedSlices(IndexedSlices(indices=Tensor("tower0/gradients/tower0/cascade_rcnn_stage2/multilevel_roi_align/concat_grad/sub_2:0", shape=(?,), dtype=int64, device=/device:GPU:0), values=Tensor("tower0/gradients/tower0/cascade_rcnn_stage2/multilevel_roi_align/concat_grad/GatherV2_8:0", shape=(?, 256, 7, 7), dtype=float32, device=/device:GPU:0), dense_shape=Tensor("tower0/gradients/tower0/cascade_rcnn_stage2/multilevel_roi_align/concat_grad/Shape_2:0", shape=(4,), dtype=int32, device=/device:GPU:0))) to a dense Tensor of unknown shape. This may consume a large amount of memory.


Can you suggest something to reduce memory consumption

**Comment by AlokRaj1994** [comment]
Should i create a customdataset.py file in datasets/instances?

**Reply by JaMe76** [reply]
I mean this is up to you. 

I personally place my custom datasets in a separate folder outside of the project.
